fontique/src/backend/dwrite.rs

```diff
@@ -120,129 +113,270 @@ impl SystemFonts {
+            let mut collection: Option<IDWriteFontCollection> = None;
+            factory
+                .GetSystemFontCollection(&mut collection, update)
+                .ok()?;
+            let collection = collection?;
+            let factory2: IDWriteFactory2 = factory.cast().ok()?;
+            let fallback = factory2.GetSystemFontFallback().ok()?;
+            Some(Self {
+                collection,
+                fallback,
+                map_buf: vec![],
+            })
+        }
+    }
+
+    fn get(&self, index: u32) -> Option<DWriteFontFamily> {
+        unsafe {
+            self.collection
+                .GetFontFamily(index)
+                .ok()
+                .map(DWriteFontFamily)
+        }
+    }
+
+    fn family_by_name(&self, name: &str) -> Option<DWriteFontFamily> {
+        let mut index = 0;
+        let mut exists = Default::default();
+        let mut name_buf: smallvec::SmallVec<[u16; 128]> = Default::default();
+        name_buf.extend(name.encode_utf16());
+        name_buf.push(0);
+        unsafe {
+            self.collection
+                .FindFamilyName(PCWSTR::from_raw(name_buf.as_ptr()), &mut index, &mut exists)
+                .ok()?;
+        }
+        self.get(index as _)
+    }
+
+    fn families(&self) -> impl Iterator<Item = DWriteFontFamily> {
+        let this = self.collection.clone();
+        unsafe {
+            let count = this.GetFontFamilyCount();
+            (0..count).filter_map(move |index| this.GetFontFamily(index).ok().map(DWriteFontFamily))
+        }
+    }
+
+    fn family_name_for_text(&mut self, text: &str, locale: Option<&str>) -> Option<String> {
+        self.map_buf.clear();
+        self.map_buf.extend(text.encode_utf16());
+        let text_len = self.map_buf.len();
+        if let Some(locale) = locale {
+            self.map_buf.extend(locale.encode_utf16());
+        }
+        self.map_buf.push(0);
+        let (text, locale) = self.map_buf.split_at(text_len);
+        let source: IDWriteTextAnalysisSource = TextSource { text, locale }.into();
+        unsafe {
+            let mut cur_offset = 0;
+            while cur_offset < text_len {
+                let mut mapped_len = 0;
+                let mut mapped_font = None;
+                if self
+                    .fallback
+                    .MapCharacters(
+                        &source,
+                        cur_offset as u32,
+                        (text_len - cur_offset) as u32,
+                        &self.collection,
+                        None,
                         DWRITE_FONT_WEIGHT_REGULAR,
                         DWRITE_FONT_STYLE_NORMAL,
                         DWRITE_FONT_STRETCH_NORMAL,
-                        &mut mapped_length,
-                        &mut font,
-                        &mut scale,
+                        &mut mapped_len,
+                        &mut mapped_font,
+                        &mut 1.0,
```

Comment:
For some reason, I feel nervous about this line: `&mut 1.0` especially since it is an `[out]` parameter. (It doesn't get used ... should it be? But the old code didn't use it either.)

Reply:
We don’t need this value nor do we really need `mapped_len` which could be passed in the same way.

If passing a mutable reference to a literal seems awkward, I can add a temporary but I happen to like this little quirk of Rust :)

Reply:
I'm fine with leaving it. Just the disadvantage of having spent so many (so many) years in C / C++. :)